pkg/kv/kvserver/store.go

```diff
@@ -411,8 +411,277 @@ func (rs *storeReplicaVisitor) EstimatedCount() int {
+
+- each Replica (and its surrounding Store) needs to run a considerable amount
+of logic to update its state when switching to the new descriptor, and now
+this logic has to be triggered by the intent resolution. This is a lot less
+convenient than using instead the roachpb.EndTxnRequest as a direct trigger
+due to the need to somehow pass the data through the intent resolution
+process.
+
+Thanks to the invariant, a Replica can switch to the new RangeDescriptor when
+it applies the corresponding EndTxnRequest.
+
+TODO(tbg): is that it? I thought there was more. Seems that even the raft
+config change stuff doesn't "really" get more involved if we didn't have
+the invariant.
+
+TODO: write something somewhere and link raft way of doing things and this: -
+https://github.com/etcd-io/etcd/issues/7625#issuecomment-489232411 Also make
+sure the comment clearly has the invariants we need for apply-time activation.
+
+INVARIANT: A Store never contains two Replicas from the same Range, nor do
+the key ranges for any two of its Replicas overlap. (Note that there
+is no requirement that these Replicas come from a consistent set of Ranges).
+
+To illustrate this last invariant, consider a split of a Replica [a-z) into two
+Replicas, [a-c) and [c,z). The Store will never contain both; it has to swap
+directly from [a-z) to [a,c)+[c,z). For a similar example, consider accepting
+a new Replica [b,d) when a Replica [a,c) is already present, which is similarly
+not allowed. To understand how such situations could arise, consider that
+a series of changes to the RangeDescriptor is observed asynchronously by
+followers. This means that a follower may have Replica [a-z) while any number
+of splits and replication changes are already known to the leaseholder, and
+the new Ranges created in the process may be attempting to add a Replica to
+the slow follower.
+
+With the invariant, we effectively allow looking up a unique (if it exists)
+Replica for any given key, and ensure that no two Replicas on a Store operate
+on shared keyspace (as seen by the storage.Engine). Refer to the Replica
+Lifecycle diagram below for details on how this invariant is upheld.
+
+Replica Lifecycle
+
+A Replica should be thought of primarily as a State Machine applying commands
+from a replicated log (the log being replicated across the members of the
+Range). The Store's RaftTransport receives Raft messages from Replicas
+residing on other Stores and routes them to the appropriate Replicas via
+Store.HandleRaftRequest (which is part of the RaftMessageHandler interface),
+ultimately resulting in a call to Replica.handleRaftReadyRaftMuLocked, which
+houses the integration with the etcd/raft library (raft.RawNode). This may
+generate Raft messages to be sent to other Stores; these are handed to
+Replica.sendRaftMessages which ultimately hands them to the Store's
+RaftTransport.SendAsync method. This is relatively straightforward unless
```

Comment:
Consider briefly spelling out here that Raft uses message passing, not request/response, and that outbound messages are always sent over an outbound gRPC/TCP connection while inbound messages are received via an inbound gRPC/TCP connection (so an asymmetric network outage is very much a possibility and happens with some frequency in the wild).

Reply:
Also: "unless when" reads a bit odd -- try either "unless" or "except when".